src/translator/swappable.cpp

```diff
@@ -0,0 +1,154 @@
+#include "marian.h"
+#include "translator/swappable.h"
+#include "common/logging.h"
+#include "data/corpus.h"
+#include "data/text_input.h"
+#include "translator/beam_search.h"
+#include "translator/translator.h"
+#include "common/io.h"
+#include "common/timer.h"
+#include <vector>
+#include "tensors/gpu/swap.h"
+
+namespace marian {
+
+void GPUEngine::SwapPointers(std::vector<MemoryPiece::PtrType> &with) {
+  auto write_it = graph_->params()->begin();
+  auto read_it = with.begin();
+  for (; read_it != with.end(); ++write_it, ++read_it) {
+    std::swap(*(*write_it)->val()->memory(), **read_it);
+  }
+}
+
+namespace {
+DeviceId LookupGPU(const Ptr<Options> options, size_t deviceIdx) {
+  auto devices = Config::getDevices(options);
+  ABORT_IF(deviceIdx >= devices.size(), "GPU device index higher than configured.");
+  return devices[deviceIdx];
+}
+} // namespace
+
+GPUEngine::GPUEngine(Ptr<Options> options, size_t deviceIdx) 
+  : options_(options), graph_(New<ExpressionGraph>()), myDeviceId_(LookupGPU(options, deviceIdx)), allocator_(myDeviceId_, 0, 128 * 1048576) {
+  ABORT_IF(myDeviceId_.type == DeviceType::cpu, "Swappable slot only works for GPU devices.");
+  options_->set("inference", true);
+  options_->set("shuffle", "none");
+
+  // Create graph
+  auto prec = options_->get<std::vector<std::string>>("precision", {"float32"});
+  graph_->setDefaultElementType(typeFromString(prec[0]));
+  graph_->setDevice(myDeviceId_);
+  graph_->reserveWorkspaceMB(options_->get<size_t>("workspace"));
+
+  scorers_ = createScorers(options_);
+  for (auto scorer : scorers_) {
+    scorer->init(graph_);
+    // TODO lexical shortlists are not supported yet.
+  }
+  graph_->forward();
+  // TODO: reach into graph_->params() private members and free the parameter memory.
+}
+
+GPUEngine::~GPUEngine() {}
+
+GPULoadedModel::GPULoadedModel(Ptr<GPUEngine> gpu) : engine_(gpu) {
+  for (auto &param : *engine_->graph_->params()) {
+    parameters_.push_back(engine_->allocator_.alloc(param->val()->memory()->size()));
+  }
+}
+
+GPULoadedModel::~GPULoadedModel() {
+  for (MemoryPiece::PtrType &p : parameters_) {
+    engine_->allocator_.free(p);
+  }
+}
+
+void GPULoadedModel::Load(const GPULoadedModel &from) {
+  srcVocabs_ = from.srcVocabs_;
+  trgVocab_ = from.trgVocab_;
+
+  ABORT_IF(engine_ != from.engine_, "TODO: copy across GPUs.");
+
+  for (size_t i = 0; i < parameters_.size(); ++i) {
+    swapper::copyGpuToGpu(reinterpret_cast<char*>(parameters_[i]->data()), reinterpret_cast<const char*>(from.parameters_[i]->data()), parameters_[i]->size(), engine_->myDeviceId_);
+  }
+}
+
+void GPULoadedModel::Load(const CPULoadedModel &from) {
+  srcVocabs_ = from.SrcVocabs();
+  trgVocab_ = from.TrgVocab();
+  for (size_t i = 0; i < parameters_.size(); ++i) {
+    swapper::copyCpuToGpu(reinterpret_cast<char*>(parameters_[i]->data()), from.Parameters()[i].data(), from.Parameters()[i].size(), engine_->myDeviceId_);
```

Comment:
@kpu could you please explain what each of the `io::Item`'s in `parameters_` corresponds to? Is it a single parameter in the .npz file? I'm trying to understand what the memory layout looks like here.

Reply:
Each of them is one tensor, corresponding to a single parameter in the npz file.  (Except for the configuration)